clippy_lints/src/utils/hir_utils.rs

```diff
@@ -574,4 +581,98 @@ impl<'a, 'tcx> SpanlessHash<'a, 'tcx> {
             },
         }
     }
+
+    pub fn hash_lifetime(&mut self, lifetime: &Lifetime) {
```

Comment:
Same as below, also hash `hir_id` and `span`. But also the `std::mem::disciminant()` of the `LifetimeName`. Also hash the disciriminant of the `ParamName` and make the match explicit by adding `Error` instead of the wildcard.